apps/api/src/app/shared/framework/idempotency.interceptor.ts

```diff
@@ -0,0 +1,188 @@
+      hash.update(Buffer.from(JSON.stringify(request.body)));
+      const bodyHash = hash.digest('hex');
+      //if 1st time we are seeing the request, marks the request as in-progress if not, does nothing
+      const isNewReq = await this.setCache(
+        cacheKey,
+        { status: ReqStatusEnum.PROGRESS, bodyHash },
+        IDEMPOTENCY_PROGRESS_TTL,
+        true
+      );
+      // Check if the idempotency key is in the cache
+      if (isNewReq) {
+        return next.handle().pipe(
+          map(async (response) => {
+            const httpResponse = context.switchToHttp().getResponse();
+            const statusCode = httpResponse.statusCode;
+
+            // Cache the success response and return it
+            await this.setCache(
+              cacheKey,
+              { status: ReqStatusEnum.SUCCESS, bodyHash, statusCode: statusCode, data: response },
+              IDEMPOTENCY_CACHE_TTL
+            );
+            Logger.verbose(`cached the success response for idempotency key:${idempotencyKey}`, LOG_CONTEXT);
+            this.setHeaders(httpResponse, { [HEADER_KEYS.IDEMPOTENCY]: idempotencyKey });
+
+            return response;
+          }),
+          catchError((err) => {
+            const httpException = this.buildError(err);
+            // Cache the error response and return it
+            const error = err instanceof HttpException ? err : httpException;
+            this.setCache(
+              cacheKey,
+              {
+                status: ReqStatusEnum.ERROR,
+                statusCode: httpException.getStatus(),
+                bodyHash,
+                data: error,
+              },
+              IDEMPOTENCY_CACHE_TTL
+            ).catch(() => {});
+            Logger.verbose(`cached the error response for idempotency`, LOG_CONTEXT);
+            this.setHeaders(context.switchToHttp().getResponse(), { [HEADER_KEYS.IDEMPOTENCY]: idempotencyKey });
+
+            return throwError(() => error);
+          })
+        );
+      } else {
+        const data = await this.cacheService.get(cacheKey);
+        this.setHeaders(context.switchToHttp().getResponse(), { [HEADER_KEYS.IDEMPOTENCY]: idempotencyKey });
+        const parsed = JSON.parse(data);
+        if (parsed.status === ReqStatusEnum.PROGRESS) {
+          // api call is in progress, so client need to handle this case
+          Logger.error(`previous api call in progress rejecting the request. key:${idempotencyKey}`, LOG_CONTEXT);
+          const error = 'IDEMPOTENCY_REQUEST_PROCESSING';
+          this.setHeaders(context.switchToHttp().getResponse(), {
+            [HEADER_KEYS.IDEMPOTENCY_CONFLICT]: error,
+            [HEADER_KEYS.RETRY_AFTER]: `1`,
+          });
+
+          return throwError(
+            () =>
+              new HttpException(
+                {
+                  error,
+                  message: `request ${idempotencyKey} is currently being processed. Please retry after 1 second`,
```

Comment:
`1 second` here should be `5 min` right? `IDEMPOTENCY_PROGRESS_TTL` is set to 5 mins but the message and `RETRY_AFTER` header say 1 second @mahendraHegde @Cliftonz

Reply:
@dammy001 In the instance that the request is still processing, we specifically having the user retry in 1 second as our average response completion time is < 1s.
This would give the server more then enough time to complete the request and the client to request the result.
So this is correct, unless I am missing something.